src/Esprima/JavascriptParser.cs

```diff
@@ -80,6 +78,54 @@ public void ReleaseLargeBuffers()
+    private static readonly Func<JavaScriptParser, Expression> parseAssignmentExpression;
+    private static readonly Func<JavaScriptParser, Expression> parseExponentiationExpression;
+    private static readonly Func<JavaScriptParser, Expression> parseUnaryExpression;
+    private static readonly Func<JavaScriptParser, Expression> parseExpression;
+    private static readonly Func<JavaScriptParser, Expression> parseNewExpression;
+    private static readonly Func<JavaScriptParser, Expression> parsePrimaryExpression;
+    private static readonly Func<JavaScriptParser, Expression> parseGroupExpression;
+    private static readonly Func<JavaScriptParser, Expression> parseArrayInitializer;
+    private static readonly Func<JavaScriptParser, Expression> parseObjectInitializer;
+    private static readonly Func<JavaScriptParser, Expression> parseBinaryExpression;
+    private static readonly Func<JavaScriptParser, Expression> parseLeftHandSideExpression;
+    private static readonly Func<JavaScriptParser, Expression> parseLeftHandSideExpressionAllowCall;
+    private static readonly Func<JavaScriptParser, Statement> parseStatement;
+    private static readonly Func<JavaScriptParser, BlockStatement> parseFunctionSourceElements;
+    private static readonly Func<JavaScriptParser, Expression> parseAsyncArgument;
+
+    static JavaScriptParser()
+    {
+        var dummyInstance = new JavaScriptParser();
+
+        parseAssignmentExpression = BuildCachedDelegateFor(dummyInstance.ParseAssignmentExpression);
+        parseExponentiationExpression = BuildCachedDelegateFor(dummyInstance.ParseExponentiationExpression);
+        parseUnaryExpression = BuildCachedDelegateFor(dummyInstance.ParseUnaryExpression);
+        parseExpression = BuildCachedDelegateFor(dummyInstance.ParseExpression);
+        parseNewExpression = BuildCachedDelegateFor(dummyInstance.ParseNewExpression);
+        parsePrimaryExpression = BuildCachedDelegateFor(dummyInstance.ParsePrimaryExpression);
+        parseGroupExpression = BuildCachedDelegateFor(dummyInstance.ParseGroupExpression);
+        parseArrayInitializer = BuildCachedDelegateFor(dummyInstance.ParseArrayInitializer);
+        parseObjectInitializer = BuildCachedDelegateFor(dummyInstance.ParseObjectInitializer);
+        parseBinaryExpression = BuildCachedDelegateFor(dummyInstance.ParseBinaryExpression);
+        parseLeftHandSideExpression = BuildCachedDelegateFor(dummyInstance.ParseLeftHandSideExpression);
+        parseLeftHandSideExpressionAllowCall = BuildCachedDelegateFor(dummyInstance.ParseLeftHandSideExpressionAllowCall);
+        parseStatement = BuildCachedDelegateFor(dummyInstance.ParseStatement);
+        parseFunctionSourceElements = BuildCachedDelegateFor(dummyInstance.ParseFunctionSourceElements);
+        parseAsyncArgument = BuildCachedDelegateFor(dummyInstance.ParseAsyncArgument);
+
+        static Func<JavaScriptParser, T> BuildCachedDelegateFor<T>(Func<T> func)
```

Comment:
this is new to me, interesting to see the perf improvements

Reply:
It's a micro-optimization (parser instances weigh 120 bytes less and initialization of delegates is a one-time cost). I don't expect much improvement from this (especially using the existing benchmarks).